### PR TITLE
README updater

### LINKED ISSUE #5679: Adapt README for release and branching
The README is used in github and is used as the long description in releases.

It works well in and is written for `master`. It is not useful on release branches `stable/2.7.x` as its still pointing to badges that reflect `master` (we don't care about development branches).

For release long description it really should reflect the static nature of the release i.e. what is the state of the requirements as specified in the release. Currently its pointing to `master` and giving misleading information.

There are some issues:
* Github has no way to dynamically change references to the branch.  So likely we need to script those change to adjust the README in critical branches.
* Some of the upstream badges and references don't quite work with point releases.

My gut feeling is that we need a script that is able to read `master` README.rst and produce a version appropriate for a branch or tag.

### PR DESCRIPTION
**Pull request by dwaynebailey**
This:

* Rewrites output from `--long-description` so that the package will point to release notes, badges, etc that correspond to the Pootle version
* Adds an `update_readme` command that will write with `-w` to README.rst to propose various README changes.

Fixes #5679

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -294,11 +294,63 @@ def adjust_installation_command():
...
                     readme_lines[ln] = "  pip install Pootle\n"
 
+    def replace_urls():
+        from pootle.core.utils import version
+        branch_escape = version.get_git_branch().replace('/', '%2F')
+        for ln, line in enumerate(readme_lines):
+            for pattern, replace, rewrite_type in (
+                # Release Notes
+                ('releases/[0-9]\.[0-9]\.[0-9]\.html', 'releases/%s.html' %
+                 version.get_main_version(), 'all'),
```
Comment (unho) [inline setup.py]: Nit: put each of the 3 values in the tuples here in a line of its own to increase readability.

File: setup.py [hunk 2]
```diff
@@ -294,11 +294,63 @@ def adjust_installation_command():
...
+                 'requires/github/translate/pootle/%s' %
+                 version.get_git_branch(), 'branch'),
+                ('requirements/\?branch=master', 'requirements/?branch=%s' %
+                 branch_escape, 'branch'),
+                ('https://img.shields.io/requires/.*',
+                 'https://requires.io/github/translate/pootle/requirements.svg?tag=%s'
+                 % version.get_version(), 'tag'),
+                ('requirements/\?branch=master', 'requirements/?tag=%s' %
+                 version.get_version(), 'tag'),
+            ):
```
Comment (unho) [inline setup.py]: Nit: put this tuple of tuples out of the `for` loops to increase code readability.

File: docs/developers/releasing.rst [hunk 3]
```diff
@@ -440,6 +441,16 @@ You can safely branch, if required, for a ``stable/`` branch before you tag.
...
 
+If you branch you will want to update the :file:`README.rst` file so that it
+points correctly to branched versions of badges and documentation.  Review and
+test the actual links created, you don't need to commit everything.
+
+.. code-block:: console
+
+    $ ./setup.py update_readme -w
+    $ git diff README.rst
+    $ git commit README.rst
```
Comment (unho) [inline docs/developers/releasing.rst]: Nit: I would add `-m "Adjust README to branch"` for faster running.

File: setup.py [hunk 4]
```diff
@@ -278,7 +310,8 @@ def parse_long_description(filename, tag=False):
 
     def reduce_header_level():
         # PyPI doesn't like title to be underlined with =
-        readme_lines[1] = readme_lines[1].replace("=", "-")
+        if tag:
+            readme_lines[1] = readme_lines[1].replace("=", "-")
```
Comment (unho) [inline setup.py]: This specific change looks like it doesn't belong to this commit.

File: setup.py [hunk 5]
```diff
@@ -200,6 +203,35 @@ def run(self):
         self.build_mo()
 
 
+class PootleUpdateReadme(Command):
+    """Process a README file for branching"""
+
+    user_options = [
+        ('write', 'w', 'Overwrite the %s file' % README_FILE),
+        ('branch', 'b', 'Rewrite using branch rewriting (default)'),
+        ('release', 'r', 'Rewrite using release or tag rewriting (default)'),
```
Comment (unho) [inline setup.py]: How can both be default? I mean, below it looks like specifying one of them alters the other (despite not being used?)...

Is it possible to simplify these options?

File: setup.py [hunk 6]
```diff
@@ -294,11 +327,77 @@ def adjust_installation_command():
                 else:
                     readme_lines[ln] = "  pip install Pootle\n"
 
+    def replace_urls():
+        from pootle.core.utils import version
+        branch = version.get_git_branch()
+        if branch is not None:
+            branch_escape = branch.replace('/', '%2F')
+        for ln, line in enumerate(readme_lines):
+            for pattern, replace, rewrite_type in (
```
Comment (unho) [inline setup.py]: Nit: put this tuple of tuples out of the `for` loops to increase code readability.

File: docs/developers/releasing.rst (hunk 3 above)
Comment (dwaynebailey) [inline docs/developers/releasing.rst]: Can we prioritise such comments based on... we'd like to get 2.8.0 out sometime this decade.  Seriously, these are releasing docs!

File: setup.py (hunk 4 above)
Comment (dwaynebailey) [inline setup.py]: Its in the right place, not dropping level is specific to tags and this commit is adding branch level changing.

File: setup.py (hunk 5 above)
Comment (dwaynebailey) [inline setup.py]: Removed the erroneous "(default)".  The rest I'm not changing.

**Comment by unho** [comment]
lgtm despite I think having the tuple of tuples at the function level and not nested in two `for` loops increases readability and allows to calculate it just once instead once per line.
